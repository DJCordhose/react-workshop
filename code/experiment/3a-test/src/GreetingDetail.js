import React, {PropTypes} from 'react';

export default class GreetingDetail extends React.Component {
    render() {
        const {name, greeting} = this.state;

        return (
            <div>
                <input ref={input => this.input = input}
                       onChange={event => this.updateModel('name', event.target.value)}
                       name="name"
                       value={name}
                       placeholder="Name"/>
                <input onChange={event => this.updateModel('greeting', event.target.value)}
                       name="greeting"
                       value={greeting}
                       placeholder="Greeting"/>
                <button
                    onClick={() => this.reset()}>
                    Clear
                </button>
                <button
                    onClick={() => this.save()}>
                    Save
                </button>
            </div>);
    }

    constructor(props) {
        super(props);
        const {name, greeting} = this.props.greeting || {name: '', greeting: ''};
        this.state = {
            name,
            greeting
        };
    }

    reset() {
        this.setState({name: '', greeting: ''});
        if (this.input) {
            this.input.focus();
        }
    }

    save() {
        const {onAdd} = this.props;
        const {name, greeting} = this.state;
        onAdd({
            name,
            greeting
        });
    }

    updateModel(name, value) {
        this.setState({[name]: value});
    }
}
GreetingDetail.propTypes = {
    greeting: PropTypes.shape({
        name: PropTypes.string.isRequired,
        greeting: PropTypes.string.isRequired
    }),
    onAdd: PropTypes.func.isRequired
};
