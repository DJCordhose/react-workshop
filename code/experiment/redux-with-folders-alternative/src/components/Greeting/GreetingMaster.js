import React from 'react';
import { connect } from 'react-redux';

import * as actions from '../../actions';
import {filterGreetings} from '../../selectors';
import {MODE_DETAIL} from '../../actions';

function GreetingMaster(props) {
    const {greetings, onAdd} = props;
    const body = greetings.map(greeting => <tr key={greeting.id}><td>{greeting.name}</td><td>{greeting.greeting}</td></tr>);
    return (
        <div>
            <table>
                <thead>
                    <tr><th>Name</th><th>Greeting</th></tr>
                </thead>
                <tbody>
                {body}
                </tbody>
            </table>
            <button className="ActionButton"
                onClick={onAdd}>
                +
            </button>
        </div>
    );
}

export default connect(
    state => ({
        greetings: filterGreetings(state.greetings, state.filter),
        mode: state.mode
    }),
    dispatch => ({
        onAdd: () => dispatch(actions.setMode(MODE_DETAIL))
    })
)(GreetingMaster);
