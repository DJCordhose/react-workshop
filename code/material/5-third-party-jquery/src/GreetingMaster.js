import React from "react";
import Spinner from "./Spinner";

const GreetingTable = ({ greetings }) => {
  return (
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Greeting</th>
        </tr>
      </thead>
      <tbody>
        {greetings.map(greeting => (
          <tr key={greeting.id}>
            <td>{greeting.name}</td>
            <td>{greeting.greeting}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default class GreetingMaster extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      limitValue: ""
    };
  }

  onLimitChange(newValue) {
    if (newValue === null || newValue === 0) {
      newValue = "";
    } else if (newValue < 1) {
      newValue = this.props.greetings.length - 1;
    } else if (newValue > this.props.greetings.length) {
      newValue = 1;
    }

    this.setState({ limitValue: newValue });
  }

  render() {
    const { greetings, onAdd } = this.props;
    const { limitValue } = this.state;

    const limitedGreetings = limitValue > 0 ? greetings.slice(0, limitValue) : greetings;

    return (
      <div>
        {/* TODO
        
        Setze hier die Spinner Komponente ein
        - Properties:
        - 'label': irgendeine Bezeichnung
        - value: limitValue aus dem State
        - onValueChange: callback-Funktion, die mit dem Ã¼bergebenen
                         neuen Wert die Methode 'this.onLimitChange'
                         aufrufen soll

        <Spinner
          . . .
        />
        
        */}

        <div>
          Showing {limitedGreetings.length} of {greetings.length} greetings
        </div>
        <GreetingTable greetings={limitedGreetings} />
        <button onClick={onAdd}>Add</button>
      </div>
    );
  }
}
