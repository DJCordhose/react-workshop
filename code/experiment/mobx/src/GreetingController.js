import React from 'react';

import {observer} from 'mobx-react';

import GreetingMaster from './GreetingMaster';
import GreetingDetail from './GreetingDetail';
import Chart from './Chart';

import {MODE_MASTER, MODE_DETAIL} from './store';

@observer
export default class GreetingController extends React.Component {
    render() {
        const {store} = this.props;
        const {mode, aggregatedGreetings, filteredGreetings} = store;
        const {loadGreeting, saveGreeting, setFilter} = store;

        return (
            <div className="Main">
                <div className="Left">
                    {mode === MODE_MASTER ?
                        <GreetingMaster greetings={filteredGreetings}
                                        onAdd={() => store.mode = MODE_DETAIL}
                        /> :
                        <GreetingDetail onAdd={(greeting) => saveGreeting(greeting)}/>
                    }
                </div>
                <div className="Right">
                    <Chart data={aggregatedGreetings} onSegmentSelected={filter => setFilter(filter)}/>
                </div>
            </div>);
    }

    componentDidMount() {
        const {loadGreetings} = this.props.store;
        loadGreetings();
    }

}

