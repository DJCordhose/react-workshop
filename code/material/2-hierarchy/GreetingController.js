import React from "react";
import GreetingMaster from "./GreetingMaster";
import GreetingDetail from "./GreetingDetail";

let currentId = 0;

const sampleGreetings = [
  {
    id: currentId++,
    name: "Olli",
    greeting: "Huhu"
  },
  {
    id: currentId++,
    name: "Oma",
    greeting: "Hallo"
  }
];

const MODE_MASTER = "MODE_MASTER";
const MODE_DETAIL = "MODE_DETAIL";

export default class GreetingController extends React.Component {
  render() {
    const { mode, greetings } = this.state;
    return (
      <div>
        {mode === MODE_MASTER ? (
          <GreetingMaster greetings={greetings} onAdd={() => this.setState({ mode: MODE_DETAIL })} />
        ) : null
        /*
                        1a. Hier statt null deine GreetingDetail-Komponente einbinden
                        1b. Eine Callback-Funktion für das Hinzufügen eines Grußes an die GreetingDetail-Komponente übergeben
                        1c. Die Callback-Funktion soll addGreeting (s.u.) aufrufen

                        2a. In GreetingDetail einen Knopf zum Hinzufügen einbauen
                        2b. Beim Drücken auf den Knopf soll die übergebene Callback-Funktion (1b) mit dem eingegebenen
                            Gruß aufgerufen werden (Parameter: ein Objekt { name: ..., greeting: ... })
                     */
        }
      </div>
    );
  }
  constructor(props) {
    super(props);
    this.state = {
      greetings: sampleGreetings,
      mode: MODE_MASTER
    };
  }

  addGreeting(greetingToBeAdded) {
    // Aktuelle Liste mit Greetings aus dem State holen
    const { greetings } = this.state;

    // dem übergebenen (neuen) Greeting eine "eindeutige" Id hinzufügen
    greetingToBeAdded.id = currentId++;

    // Ein neues Array mit den alten Greetings und dem neuen Greeting erzeugen
    const newGreetings = [...greetings, greetingToBeAdded];

    // den State mit den neuen Greetings setzen, außerdem wieder die MASTER-View anzeigen
    this.setState({
      greetings: newGreetings,
      mode: MODE_MASTER
    });
  }
}
