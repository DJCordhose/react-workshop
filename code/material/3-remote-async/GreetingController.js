import React from "react";

import GreetingMaster from "./GreetingMaster";
import GreetingDetail from "./GreetingDetail";

const BACKEND_URL = "http://localhost:7000/greetings";
const MODE_MASTER = "MODE_MASTER";
const MODE_DETAIL = "MODE_DETAIL";

export default class GreetingController extends React.Component {
  render() {
    const { mode, greetings } = this.state;
    return (
      <div>
        {mode === MODE_MASTER ? (
          <GreetingMaster greetings={greetings} onAdd={() => this.setState({ mode: MODE_DETAIL })} />
        ) : (
          <GreetingDetail onSave={greeting => this.saveGreeting(greeting)} />
        )}
      </div>
    );
  }

  constructor(props) {
    super(props);
    this.state = {
      greetings: [],
      mode: MODE_MASTER
    };
  }

  componentDidMount() {
    this.loadGreetings();
  }

  async loadGreetings() {
    let greetings;

    try {
      // TODO:
      // 1. Mit fetch Serverzugriff machen
      // 2. Aus der Antwort das JSON-Objekt auslesen (zuweisen an 'greetings')
      //   (Die gelesene Antwort entspricht unserem Greetings-Array)
    } catch (err) {
      console.error("LOADING GREETINGS FAILED:", err);
      return;
    }

    // 3. setze die gelesenen Greetings (=> Antwort vom Server)
    // als State neu
  }

  async saveGreeting(greetingToBeAdded) {
    let newGreeting;

    try {
      // TODO:
      // 1. Vervollst√§ndige den fetch-Aufruf
      fetch(BACKEND_URL, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json"
        },
        body: JSON.stringify(greetingToBeAdded)
      });

      // (1b Optional: wirf einen Error wenn der Resonse Status nicht 201 ist)

      // 2. Lies aus der Antwort das JSON Objekt aus (zuweisen an 'newGreeting')
      //    (Die Antwort entspricht dem neuen Greeting + einer Id, die auf dem
      //     Server erzeugt wurde)
    } catch (err) {
      console.error("LOADING GREETINGS FAILED:", err);
    }

    // 3. Erzeuge eine neue Liste mit Greetings
    //    (bestehende Liste aus dem State ('greetings') und das eben
    //    gelesene Greeting ('newGreeting'))

    // 4. Setze den Zustand neu:
    //    - 'greetings' mit den neuen Greetings
    //    - 'mode' auf MODE_MASTER, um den Master-View wieder anzuzeigen
  }
}
