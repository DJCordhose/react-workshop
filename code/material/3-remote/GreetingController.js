import React from "react";

import GreetingMaster from "./GreetingMaster";
import GreetingDetail from "./GreetingDetail";

const BACKEND_URL = "http://localhost:7000/greetings";
const MODE_MASTER = "MODE_MASTER";
const MODE_DETAIL = "MODE_DETAIL";

export default class GreetingController extends React.Component {
  render() {
    const { mode, greetings } = this.state;
    return (
      <div>
        {mode === MODE_MASTER ? (
          <GreetingMaster greetings={greetings} onAdd={() => this.setState({ mode: MODE_DETAIL })} />
        ) : (
          <GreetingDetail onSave={greeting => this.saveGreeting(greeting)} />
        )}
      </div>
    );
  }

  constructor(props) {
    super(props);
    this.state = {
      greetings: [],
      mode: MODE_MASTER
    };
  }

  componentDidMount() {
    this.loadGreetings();
  }

  loadGreetings() {
    return (
      fetch(BACKEND_URL)
        /* TODO:

            1. Hole das JSON-Objekt aus der Server-Response
            2. Das JSON Objekt entspricht unserem Greetings-Array. Setze den State damit neu
             */
        .catch(err => console.error("LOADING GREETINGS FAILED:", err))
    );
  }

  saveGreeting(greetingToBeAdded) {
    fetch(BACKEND_URL, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json"
      },
      body: JSON.stringify(greetingToBeAdded)
    })
      .then(response => {
        /* TODO: Liefer das JSON-Objekt aus der Antwort zurück

            (Optional: Wenn das Status Code NICHT 201 ist, wirf einen Error
             */
      })
      .then(json => {
        /* TODO: Das JSON-Objekt, das wir vom Server erhalten haben
               enthält als einziges das Property 'id', das die vom Server
               vergebenen Id für unser neues Greeting darstellt.

               1. Baue ein neues Greeting-Objekt mit den Daten aus
               greetingToBeAdded und der Id aus dem JSON-Objekt

               2. Erzeuge eine neue Greetings-Liste mit den bisherigen
               Greetings und dem neuen Greeting, dass Du in Schritt 2
               erzeugt hast

               3. Setze die erzeugte Greetings-Liste als neuen State
             */
      });
  }
}
