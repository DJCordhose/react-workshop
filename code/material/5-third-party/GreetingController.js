import React from 'react';

import {loadFromServer, saveToServer} from './backend';

import GreetingMaster from './GreetingMaster';
import GreetingDetail from './GreetingDetail';
import Chart from './Chart';
import {aggregateGreetings} from './util';

const MODE_MASTER = 'MODE_MASTER';
const MODE_DETAIL = 'MODE_DETAIL';

export default class GreetingController extends React.Component {
    render() {
        const {mode, greetings, filter} = this.state;
        const aggregatedGreetings = aggregateGreetings(greetings);
        const filtered = filter ? greetings.filter(greeting => greeting.name === filter) : greetings;

        return (
            <div className="Main">
                <div className="Left">
                    {mode === MODE_MASTER ?
                        <GreetingMaster greetings={filtered}
                                        onAdd={() => this.setState({mode: MODE_DETAIL})}
                        /> :
                        <GreetingDetail onAdd={(greeting) => this.saveGreeting(greeting)}/>
                    }
                </div>
                <div className="Right">
                    {

                        // Hier fehlt die Einbindung des Charts:
                        // 1. du musst die bereits in Zeile 16 berechnete Aggregation der Grüße übergeben
                        // 2. Du musst ein Callback übergeben, der - wenn aufgerufen - mit dem Namen des Segments den Filter aktualisiert
                        // 3. Der Filter ist bereits im Zustand vorgesehen und mit null initialisiert
                    }
                </div>
            </div>);
    }

    constructor(props) {
        super(props);
        this.state = {
            greetings: [],
            mode: MODE_MASTER,
            filter: null
        };
    }

    componentDidMount() {
        this.loadGreetings();
    }

    loadGreetings() {
        loadFromServer(
            greetings => this.setState({greetings}),
            err => console.error('LOADING GREETINGS FAILED:', err)
        );
    }


    saveGreeting(greetingToBeAdded) {
        const _addNewGreeting = serverResponse => {
            // the server responded with the id of the new Greeting
            const newGreetingId = serverResponse.id;
            // create a new Greeting object that contains the received id
            // (create a new object for immutability)
            const newGreeting = Object.assign({}, greetingToBeAdded, {id: newGreetingId});
            // add the new greetings to the list of all greetings
            // (create a new array for immutability)
            const newGreetings = this.state.greetings.concat(newGreeting);

            // set the new list of greetings as our new state
            // also set 'MODE_MASTER' to make sure the master-View is
            // displayed now
            this.setState({
                greetings: newGreetings,
                mode: MODE_MASTER
            });

            return newGreeting;
        };

        const _reportError = err => console.error('COULD NOT SAVE GREETING: ', err);

        saveToServer(greetingToBeAdded, _addNewGreeting, _reportError);
    }
}

