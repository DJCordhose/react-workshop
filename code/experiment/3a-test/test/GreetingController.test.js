import React from 'react';
import GreetingMaster from '../src/GreetingMaster';
import GreetingDetail from '../src/GreetingDetail';
import GreetingController from '../src/GreetingController';
import backend from '../src/backend';
import {shallow, mount} from 'enzyme';

jest.mock('../src/backend', () => ({
    loadFromServer: jest.fn((success, failure) => success([
        {id: 1, name: 'Klaus', greeting: 'Moin moin'},
        {id: 2, name: 'Susi', greeting: 'Hello!'}
    ]))
}));

test('it should render master view by default', () => {
    const component = shallow(<GreetingController/>);
    expect(component).toMatchSnapshot();
});

// TODO: Write more tests... but: is difficult due to
// the internal structure of the controller component

test('it should render greetings received from server', () => {
    const component = mount(<GreetingController  />);
    expect(backend.loadFromServer.mock.calls).toHaveLength(1);

    expect(component.find('button')).toHaveLength(1);
    expect(component.find('tbody > tr')).toHaveLength(2);
});

