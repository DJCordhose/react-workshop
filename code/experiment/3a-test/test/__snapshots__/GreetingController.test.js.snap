// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render master view by default 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <GreetingMaster
        greetings={Array []}
        onAdd={[Function]}
    />
</div>,
  "nodes": Array [
    <div>
      <GreetingMaster
            greetings={Array []}
            onAdd={[Function]}
      />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <GreetingController />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": GreetingController {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "greetings": Array [],
          "mode": "MODE_MASTER",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <GreetingMaster
                    greetings={Array []}
                    onAdd={[Function]}
          />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <GreetingMaster
                    greetings={Array []}
                    onAdd={[Function]}
          />
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <GreetingController />,
}
`;
